Sources/PackageCollections/Storage/PackageCollectionsProfileStorage.swift

```diff
@@ -0,0 +1,339 @@
+            } catch {
+                callback(.failure(error))
+            }
+        }
+    }
+
+    func listSources(in profile: PackageCollectionsModel.Profile, callback: @escaping (Result<[PackageCollectionsModel.PackageCollectionSource], Error>) -> Void) {
+        self.queue.async {
+            do {
+                let profiles = try self.lock.withLock {
+                    try self.loadFromDisk()
+                }
+                callback(.success(profiles[profile] ?? []))
+            } catch {
+                callback(.failure(error))
+            }
+        }
+    }
+
+    func add(source: PackageCollectionsModel.PackageCollectionSource,
+             order: Int?,
+             to profile: PackageCollectionsModel.Profile,
+             callback: @escaping (Result<Void, Error>) -> Void) {
+        self.queue.async {
+            do {
+                try self.lock.withLock {
+                    var profiles = try self.loadFromDisk()
+                    var sources = profiles[profile]?.filter { $0 != source } ?? []
+                    let order = order.flatMap { $0 >= 0 && $0 < sources.endIndex ? order : sources.endIndex } ?? sources.endIndex
+                    sources.insert(source, at: order)
+                    profiles[profile] = sources
+                    try self.saveToDisk(profiles)
+                }
+                callback(.success(()))
+            } catch {
+                callback(.failure(error))
+            }
+        }
+    }
+
+    func remove(source: PackageCollectionsModel.PackageCollectionSource,
+                from profile: PackageCollectionsModel.Profile,
+                callback: @escaping (Result<Void, Error>) -> Void) {
+        self.queue.async {
+            do {
+                try self.lock.withLock {
+                    var profiles = try self.loadFromDisk()
+                    guard let sources = profiles[profile] else {
+                        throw Errors.invalidProfile(profile)
+                    }
+                    profiles[profile] = sources.filter { $0 != source }
+                    try self.saveToDisk(profiles)
+                }
+                callback(.success(()))
+            } catch {
+                callback(.failure(error))
+            }
+        }
+    }
+
+    func move(source: PackageCollectionsModel.PackageCollectionSource,
+              to order: Int,
+              in profile: PackageCollectionsModel.Profile,
+              callback: @escaping (Result<Void, Error>) -> Void) {
+        self.queue.async {
+            do {
+                try self.lock.withLock {
+                    var profiles = try self.loadFromDisk()
+                    guard var sources = profiles[profile] else {
+                        throw Errors.invalidProfile(profile)
+                    }
+                    sources = sources.filter { $0 != source }
+                    let order = order >= 0 && order < sources.endIndex ? order : sources.endIndex
```

Comment:
Should this return an Error if the index is out-of-bounds, similar to what subscript accessors do?  Or should this be more lenient?

Reply:
@abertelrud I think lenient is fine